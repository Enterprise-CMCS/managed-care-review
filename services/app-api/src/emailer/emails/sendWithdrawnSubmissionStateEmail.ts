import type { ContractType } from '../../domain-models'
import type { RateForDisplay } from '../../postgres/contractAndRates/withdrawContract'
import type { EmailConfiguration, EmailData } from '../emailer'
import { pruneDuplicateEmails } from '../formatters'
import { renderTemplate, stripHTMLFromTemplate } from '../templateHelpers'
import { parseAndValidateContractsAndRates } from './sendWithdrawnSubmissionCMSEmail'

export const sendWithdrawnSubmissionStateEmail = async (
    withdrawnContract: ContractType,
    ratesForDisplay: RateForDisplay[],
    config: EmailConfiguration
): Promise<EmailData | Error> => {
    const stateContactEmails: string[] = []
    const contractRev = withdrawnContract.packageSubmissions[0].contractRevision
    contractRev.formData.stateContacts.forEach((contact) => {
        if (contact.email) stateContactEmails.push(contact.email)
    })

    const toAddresses = pruneDuplicateEmails([
        ...stateContactEmails,
        ...config.devReviewTeamEmails,
    ])

    const etaData = parseAndValidateContractsAndRates(
        withdrawnContract,
        ratesForDisplay,
        config
    )
    if (etaData instanceof Error) {
        return etaData
    }

    const template = await renderTemplate(
        'sendWithdrawnSubmissionStateEmail',
        etaData
    )

    if (template instanceof Error) {
        return template
    } else {
        return {
            toAddresses,
            replyToAddresses: [],
            sourceEmail: config.emailSource,
            subject: `${
                config.stage !== 'prod' ? `[${config.stage}] ` : ''
            }${etaData.contractName} was withdrawn by CMS`,
            bodyText: stripHTMLFromTemplate(template),
            bodyHTML: template,
        }
    }
}
