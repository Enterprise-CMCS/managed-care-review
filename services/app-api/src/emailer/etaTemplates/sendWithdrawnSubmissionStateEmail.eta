<h1 style=<%~'"font-weight: normal;"'%>>Submission <%= it.contractName %> was withdrawn by CMS</h1>
<b>Withdrawn by:</b> <%= it.updatedBy %><br />
<b>Withdrawn on:</b> <%= it.updatedAt %><br />
<b>Reason for withdrawal:</b> <%= it.reason %><br />
<% if (it.formattedRates && it.formattedRates.length > 0) { %>
<b>Related rate certification(s):</b>
<ul>
<% it.formattedRates.forEach(function(rate) { %>
  <li><%= rate.rateCertificationName %></li>
<% }) %>
</ul>
<% } %>
<% if (it.formattedRates.length === 0) { %>
<br />
<% } %>
<a href="<%= it.contractSummaryURL %>">View submission in MC-Review</a>
