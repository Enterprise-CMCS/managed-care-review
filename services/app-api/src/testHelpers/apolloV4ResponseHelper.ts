import type { GraphQLFormattedError } from 'graphql'
import type { ApolloServer } from '@apollo/server'

export type TestGraphQLResponse<T = any> = {
    data?: T
    errors?: GraphQLFormattedError[]
}

export function extractGraphQLResponse<T = any>(
    response: any
): TestGraphQLResponse<T> {
    if ('body' in response && response.body) {
        const result =
            response.body.kind === 'single'
                ? response.body.singleResult
                : response.body

        return {
            data: result.data,
            errors: result.errors,
        }
    }

    return {
        data: response.data,
        errors: response.errors,
    }
}

export async function executeGraphQLOperation<TData = any, TVariables = any>(
    server: ApolloServer,
    options: {
        query: any
        variables?: TVariables
        contextValue?: any
    }
): Promise<TestGraphQLResponse<TData>> {
    const response = await server.executeOperation({
        query: options.query,
        variables: options.variables as any,
    })

    return extractGraphQLResponse<TData>(response)
}
