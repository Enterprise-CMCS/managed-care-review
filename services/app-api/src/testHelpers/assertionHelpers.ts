// Functions that assists with type narrow in jest tests

// must -  interrupts test flow to through an error
function must<T>(maybeErr: T | Error): T {
    if (maybeErr instanceof Error) {
        throw maybeErr
    }
    return maybeErr
}

// expectToBeDefined - properly  type narrows the results of toBeDefined
// https://github.com/DefinitelyTyped/DefinitelyTyped/issues/41179 would eventually address this but is not implemented
function expectToBeDefined<T>(arg: T): asserts arg is NonNullable<T> {
    expect(arg).toBeDefined()
}

export { must, expectToBeDefined }
