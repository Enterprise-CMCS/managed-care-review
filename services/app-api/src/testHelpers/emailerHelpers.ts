import type { EmailConfiguration, EmailData, Emailer } from '../emailer'
import { emailer } from '../emailer'
import type { ProgramArgType } from '@mc-review/submissions'
import type {
    ContractRevisionType,
    ContractType,
    ContractQuestionType,
    RatePackageSubmissionType,
    RateRevisionType,
    RateType,
    UnlockedContractType,
    UpdateInfoType,
    RateQuestionType,
} from '../domain-models'
import { SESServiceException } from '@aws-sdk/client-ses'
import { testSendSESEmail } from './awsSESHelpers'
import { testCMSUser, testStateUser } from './userHelpers'
import { v4 as uuidv4 } from 'uuid'
import { constructTestEmailer } from './gqlHelpers'
import type { Store } from '../postgres'

const testEmailConfig = (): EmailConfiguration => ({
    stage: 'LOCAL',
    baseUrl: 'http://localhost',
    emailSource: 'emailSource@example.com',
    devReviewTeamEmails: ['devreview1@example.com', 'devreview2@example.com'],
    cmsReviewHelpEmailAddress: '"MCOG Example" <mcog@example.com>',
    cmsRateHelpEmailAddress: '"Rates Example" <rates@example.com>',
    oactEmails: ['ratesreview@example.com'],
    dmcpReviewEmails: ['policyreview1@example.com'],
    dmcpSubmissionEmails: ['policyreviewsubmission1@example.com'],
    dmcoEmails: ['overallreview@example.com'],
    helpDeskEmail: '"MC-Review Help Desk" <MC_Review_HelpDesk@example.com>',
})

const testDuplicateEmailConfig: EmailConfiguration = {
    stage: 'LOCAL',
    baseUrl: 'http://localhost',
    emailSource: 'emailSource@example.com',
    devReviewTeamEmails: [
        'duplicate@example.com',
        'duplicate@example.com',
        'duplicate@example.com',
    ],
    cmsReviewHelpEmailAddress: 'duplicate@example.com',
    cmsRateHelpEmailAddress: 'duplicate@example.com',
    oactEmails: ['duplicate@example.com', 'duplicate@example.com'],
    dmcpReviewEmails: ['duplicate@example.com', 'duplicate@example.com'],
    dmcpSubmissionEmails: ['duplicate@example.com', 'duplicate@example.com'],
    dmcoEmails: ['duplicate@example.com', 'duplicate@example.com'],
    helpDeskEmail: 'duplicate@example.com',
}

const testStateAnalystsEmails: string[] = [
    '"State Analyst 1" <StateAnalyst1@example.com>',
    '"State Analyst 2" <StateAnalyst2@example.com>',
]

const testDuplicateStateAnalystsEmails: string[] = [
    'duplicate@example.com',
    'duplicate@example.com',
]
const s3DlUrl =
    'https://fake-bucket.s3.amazonaws.com/file.pdf?AWSAccessKeyId=AKIAIOSFODNN7EXAMPLE&Expires=1719564800&Signature=abc123def456ghijk' //pragma: allowlist secret

const sendTestEmails = async (emailData: EmailData): Promise<void | Error> => {
    try {
        await testSendSESEmail(emailData)
    } catch (err) {
        if (err instanceof SESServiceException) {
            return new Error(
                'SES email send failed. Error is from Amazon SES. Error: ' +
                    JSON.stringify(err)
            )
        }
        return new Error('SES email send failed. Error: ' + err)
    }
}

function testEmailer(customConfig?: EmailConfiguration): Emailer {
    const config = customConfig || testEmailConfig()
    return emailer(config, vi.fn(sendTestEmails))
}

async function testEmailerFromDatabase(
    store: Store,
    customConfig?: EmailConfiguration
): Promise<Emailer> {
    const mockEmailer = await constructTestEmailer({
        emailConfig: customConfig ?? testEmailConfig(),
        postgresStore: store,
    })

    mockEmailer.sendEmail = vi.fn()

    return mockEmailer
}

type State = {
    name: string
    programs: ProgramArgType[]
    code: string
}

export function mockMNState(): State {
    return {
        name: 'Minnesota',
        programs: [
            {
                id: 'abbdf9b0-c49e-4c4c-bb6f-040cb7b51cce',
                fullName: 'Special Needs Basic Care',
                name: 'SNBC',
                isRateProgram: false,
            },
            {
                id: 'd95394e5-44d1-45df-8151-1cc1ee66f100',
                fullName: 'Prepaid Medical Assistance Program',
                name: 'PMAP',
                isRateProgram: false,
            },
            {
                id: 'ea16a6c0-5fc6-4df8-adac-c627e76660ab',
                fullName: 'Minnesota Senior Care Plus ',
                name: 'MSC+',
                isRateProgram: false,
            },
            {
                id: '3fd36500-bf2c-47bc-80e8-e7aa417184c5',
                fullName: 'Minnesota Senior Health Options',
                name: 'MSHO',
                isRateProgram: false,
            },
        ],
        code: 'MN',
    }
}

export function mockMSState(): State {
    return {
        name: 'Mississippi',
        programs: [
            {
                id: 'e0819153-5894-4153-937e-aad00ab01a8f',
                fullName: 'Mississippi Coordinated Access Network',
                name: 'MSCAN',
                isRateProgram: false,
            },
            {
                id: '36c54daf-7611-4a15-8c3b-cdeb3fd7e25a',
                fullName: 'CHIP',
                name: 'CHIP',
                isRateProgram: false,
            },
        ],
        code: 'MS',
    }
}

const mockContractRev = (
    submissionPartial?: Partial<ContractRevisionType>
): ContractRevisionType => {
    return {
        createdAt: new Date('01/01/2021'),
        updatedAt: new Date('02/01/2021'),
        contract: {
            stateCode: 'MN',
            stateNumber: 3,
            id: '12345',
        },
        id: 'test-abc-125',
        formData: {
            programIDs: ['abbdf9b0-c49e-4c4c-bb6f-040cb7b51cce'],
            populationCovered: 'MEDICAID',
            submissionType: 'CONTRACT_AND_RATES',
            riskBasedContract: false,
            submissionDescription: 'A submitted submission',
            stateContacts: [
                {
                    name: 'Test Person',
                    titleRole: 'A Role',
                    email: 'test+state+contact@example.com',
                },
            ],
            supportingDocuments: [
                {
                    s3URL: 's3://bucketname/key/test1',
                    name: 'foo',
                    sha256: 'fakesha',
                    dateAdded: new Date('02/01/2021'),
                },
            ],
            contractType: 'BASE',
            contractExecutionStatus: undefined,
            contractDocuments: [
                {
                    s3URL: 's3://bucketname/key/test1',
                    name: 'foo',
                    sha256: 'fakesha',
                    dateAdded: new Date('02/01/2021'),
                },
            ],
            contractDateStart: new Date('01/01/2024'),
            contractDateEnd: new Date('01/01/2025'),
            managedCareEntities: ['MCO'],
            federalAuthorities: ['VOLUNTARY', 'BENCHMARK'],
            inLieuServicesAndSettings: undefined,
            modifiedBenefitsProvided: undefined,
            modifiedGeoAreaServed: undefined,
            modifiedMedicaidBeneficiaries: undefined,
            modifiedRiskSharingStrategy: undefined,
            modifiedIncentiveArrangements: undefined,
            modifiedWitholdAgreements: undefined,
            modifiedStateDirectedPayments: undefined,
            modifiedPassThroughPayments: undefined,
            modifiedPaymentsForMentalDiseaseInstitutions: undefined,
            modifiedMedicalLossRatioStandards: undefined,
            modifiedOtherFinancialPaymentIncentive: undefined,
            modifiedEnrollmentProcess: undefined,
            modifiedGrevienceAndAppeal: undefined,
            modifiedNetworkAdequacyStandards: undefined,
            modifiedLengthOfContract: undefined,
            modifiedNonRiskPaymentArrangements: undefined,
            statutoryRegulatoryAttestation: undefined,
            statutoryRegulatoryAttestationDescription: undefined,
        },
        ...submissionPartial,
    }
}

const mockRateRevision = (
    rateRevPartial?: Partial<RateRevisionType>
): RateRevisionType => {
    return {
        id: '12345',
        rateID: '6789',
        submitInfo: undefined,
        unlockInfo: undefined,
        createdAt: new Date(11 / 27 / 2023),
        updatedAt: new Date(11 / 27 / 2023),
        formData: {
            id: 'test-id-1234',
            rateID: 'test-id-1234',
            rateType: 'NEW',
            rateCapitationType: 'RATE_CELL',
            rateDocuments: [
                {
                    s3URL: 's3://bucketname/key/test1',
                    name: 'foo',
                    sha256: 'fakesha',
                    dateAdded: new Date(11 / 27 / 2023),
                },
            ],
            supportingDocuments: [],
            rateDateStart: new Date('01/01/2024'),
            rateDateEnd: new Date('01/01/2025'),
            rateDateCertified: new Date('01/01/2024'),
            amendmentEffectiveDateStart: new Date('01/01/2024'),
            amendmentEffectiveDateEnd: new Date('01/01/2025'),
            rateProgramIDs: ['3fd36500-bf2c-47bc-80e8-e7aa417184c5'],
            rateCertificationName: 'Rate Cert Name',
            certifyingActuaryContacts: [
                {
                    actuarialFirm: 'DELOITTE',
                    name: 'Actuary Contact 1',
                    titleRole: 'Test Actuary Contact 1',
                    email: 'actuarycontact1@example.com',
                },
            ],
            addtlActuaryContacts: [],
            actuaryCommunicationPreference: 'OACT_TO_ACTUARY',
            packagesWithSharedRateCerts: [
                {
                    packageName: 'pkgName',
                    packageId: '12345',
                    packageStatus: 'SUBMITTED',
                },
            ],
        },
        ...rateRevPartial,
    }
}

const mockRate = (ratePartial?: Partial<RateType>): RateType => {
    const submitInfo: UpdateInfoType = {
        updatedAt: new Date('02/01/2021'),
        updatedBy: {
            role: 'STATE_USER',
            email: 'someone@example.com',
            givenName: 'Someone',
            familyName: 'Example',
        },
        updatedReason: 'Initial submission',
    }
    const rateRev = mockRateRevision({
        submitInfo,
    })
    const contractRev = mockContractRev({
        id: 'test-contract-234',
        submitInfo,
    })
    const rateSubmission: RatePackageSubmissionType = {
        submitInfo,
        submittedRevisions: [rateRev, contractRev],
        rateRevision: rateRev,
        contractRevisions: [contractRev],
    }

    return {
        id: 'test-rate-234',
        createdAt: new Date('01/01/2021'),
        updatedAt: new Date('02/01/2021'),
        status: 'SUBMITTED',
        stateCode: 'MN',
        stateNumber: 2,
        parentContractID: 'test-contract-234',
        reviewStatus: 'UNDER_REVIEW',
        consolidatedStatus: 'SUBMITTED',
        revisions: [rateRev],
        packageSubmissions: [rateSubmission],
        questions: {
            DMCOQuestions: {
                totalCount: 1,
                edges: [
                    {
                        node: mockRateQuestionAndResponses({
                            id: 'dmco-rate-question-1',
                            rateID: 'test-rate-234',
                            createdAt: new Date('2024-04-23'),
                            addedBy: testCMSUser({
                                divisionAssignment: 'DMCO',
                            }),
                            division: 'DMCO',
                            documents: [
                                {
                                    id: 'dmco-rate-question-1',
                                    name: 'dmco-rate-question-1',
                                    s3URL: 's3://bucketName/key/dmco-rate-question-1-doc',
                                    downloadURL:
                                        'https://fake-bucket.s3.amazonaws.com/test',
                                },
                            ],
                        }),
                    },
                ],
            },
            DMCPQuestions: {
                totalCount: 0,
                edges: [],
            },
            OACTQuestions: {
                totalCount: 0,
                edges: [],
            },
        },

        ...ratePartial,
    }
}

const mockUnlockedContract = (
    contractPartial?: Partial<UnlockedContractType>,
    rateParitals?: Partial<RateType>[]
): UnlockedContractType => {
    const draftRates = rateParitals
        ? rateParitals.map((r) => mockRate(r))
        : [mockRate()]

    return {
        ...mockContract(),
        id: 'test-contract-123',
        createdAt: new Date('01/01/2021'),
        updatedAt: new Date('02/01/2021'),
        status: 'UNLOCKED',
        consolidatedStatus: 'UNLOCKED',
        reviewStatus: 'UNDER_REVIEW',
        stateCode: 'MN',
        stateNumber: 4,
        draftRevision: mockContractRev(),
        draftRates,
        ...contractPartial,
    }
}

const mockContract = (
    contractPartial?: Partial<ContractType>
): ContractType => {
    return {
        id: 'test-contract-123',
        contractSubmissionType: 'HEALTH_PLAN',
        createdAt: new Date('01/01/2021'),
        updatedAt: new Date('02/01/2021'),
        status: 'SUBMITTED',
        reviewStatus: 'UNDER_REVIEW',
        consolidatedStatus: 'SUBMITTED',
        stateCode: 'MN',
        stateNumber: 4,
        revisions: [],
        packageSubmissions: [
            {
                submitInfo: {
                    updatedAt: new Date('01/01/2023'),
                    updatedBy: {
                        email: 'example@example.com',
                        role: 'STATE_USER',
                        givenName: 'John',
                        familyName: 'Vila',
                    },
                    updatedReason: 'contract submit',
                },
                submittedRevisions: [],
                contractRevision: {
                    createdAt: new Date('01/01/2024'),
                    updatedAt: new Date('01/01/2025'),
                    id: '123',
                    contract: {
                        id: 'test-abc-123',
                        stateCode: 'MN',
                        stateNumber: 4,
                    },
                    submitInfo: {
                        updatedAt: new Date(),
                        updatedBy: {
                            email: 'example@example.com',
                            role: 'STATE_USER',
                            givenName: 'John',
                            familyName: 'Vila',
                        },
                        updatedReason: 'contract submit',
                    },
                    unlockInfo: undefined,
                    formData: {
                        programIDs: ['abbdf9b0-c49e-4c4c-bb6f-040cb7b51cce'],
                        populationCovered: 'MEDICAID',
                        submissionType: 'CONTRACT_AND_RATES',
                        riskBasedContract: true,
                        submissionDescription: 'A real submission',
                        supportingDocuments: [
                            {
                                s3URL: 's3://bucketname/key/contractsupporting1',
                                sha256: 'fakesha',
                                name: 'contractSupporting1',
                                dateAdded: new Date('01/15/2024'),
                                downloadURL: s3DlUrl,
                            },
                            {
                                s3URL: 's3://bucketname/key/contractSupporting2',
                                sha256: 'fakesha',
                                name: 'contractSupporting2',
                                dateAdded: new Date('01/13/2024'),
                                downloadURL: s3DlUrl,
                            },
                        ],
                        stateContacts: [
                            {
                                name: 'contract-state-contact',
                                titleRole: 'state contact',
                                email: 'contract-state-contact@example.com',
                            },
                        ],
                        contractType: 'AMENDMENT',
                        contractExecutionStatus: 'EXECUTED',
                        contractDocuments: [
                            {
                                s3URL: 's3://bucketname/key/contract',
                                sha256: 'fakesha',
                                name: 'contract',
                                dateAdded: new Date('01/01/2024'),
                                downloadURL: s3DlUrl,
                            },
                        ],
                        contractDateStart: new Date(),
                        contractDateEnd: new Date(),
                        managedCareEntities: ['MCO'],
                        federalAuthorities: ['STATE_PLAN'],
                        inLieuServicesAndSettings: true,
                        modifiedBenefitsProvided: true,
                        modifiedGeoAreaServed: false,
                        modifiedMedicaidBeneficiaries: true,
                        modifiedRiskSharingStrategy: true,
                        modifiedIncentiveArrangements: false,
                        modifiedWitholdAgreements: false,
                        modifiedStateDirectedPayments: true,
                        modifiedPassThroughPayments: true,
                        modifiedPaymentsForMentalDiseaseInstitutions: false,
                        modifiedMedicalLossRatioStandards: true,
                        modifiedOtherFinancialPaymentIncentive: false,
                        modifiedEnrollmentProcess: true,
                        modifiedGrevienceAndAppeal: false,
                        modifiedNetworkAdequacyStandards: true,
                        modifiedLengthOfContract: false,
                        modifiedNonRiskPaymentArrangements: true,
                        statutoryRegulatoryAttestation: true,
                        statutoryRegulatoryAttestationDescription:
                            'everything meets regulatory attestation',
                    },
                },
                rateRevisions: [
                    {
                        id: '1234',
                        rateID: '123',
                        createdAt: new Date('01/01/2023'),
                        updatedAt: new Date('01/01/2023'),
                        formData: {
                            rateCertificationName: 'rate cert',
                            rateType: 'AMENDMENT',
                            rateCapitationType: 'RATE_CELL',
                            rateDocuments: [
                                {
                                    s3URL: 's3://bucketname/key/rate',
                                    sha256: 'fakesha',
                                    name: 'rate',
                                    dateAdded: new Date('01/01/2023'),
                                    downloadURL: s3DlUrl,
                                },
                            ],
                            supportingDocuments: [
                                {
                                    s3URL: 's3://bucketname/key/rateSupporting1',
                                    sha256: 'fakesha',
                                    name: 'rate supporting 1',
                                    dateAdded: new Date('01/15/2023'),
                                    downloadURL: s3DlUrl,
                                },
                                {
                                    s3URL: 's3://bucketname/key/rateSupporting1',
                                    sha256: 'fakesha',
                                    name: 'rate supporting 2',
                                    dateAdded: new Date('01/15/2023'),
                                    downloadURL: s3DlUrl,
                                },
                            ],
                            rateDateStart: new Date(),
                            rateDateEnd: new Date(),
                            rateDateCertified: new Date(),
                            amendmentEffectiveDateStart: new Date(),
                            amendmentEffectiveDateEnd: new Date(),
                            rateProgramIDs: [
                                'abbdf9b0-c49e-4c4c-bb6f-040cb7b51cce',
                            ],
                            deprecatedRateProgramIDs: [],
                            certifyingActuaryContacts: [
                                {
                                    actuarialFirm: 'DELOITTE',
                                    name: 'Actuary Contact 1',
                                    titleRole: 'Test Actuary Contact 1',
                                    email: 'actuarycontact1@test.com',
                                },
                            ],
                            addtlActuaryContacts: [
                                {
                                    actuarialFirm: 'DELOITTE',
                                    name: 'Actuary Contact 1',
                                    titleRole: 'Test Actuary Contact 1',
                                    email: 'actuarycontact1@test.com',
                                },
                            ],
                            actuaryCommunicationPreference: 'OACT_TO_ACTUARY',
                            packagesWithSharedRateCerts: [],
                        },
                    },
                ],
            },
        ],
        ...contractPartial,
    }
}

const mockCommonQuestionAndResponses = <
    T extends
        | (Partial<ContractQuestionType> & { contractID: string })
        | (Partial<RateQuestionType> & { rateID: string }),
    R extends ContractQuestionType | RateQuestionType,
>(
    questionData?: T
): R => {
    const question = {
        id: `test-question-id-1`,
        createdAt: new Date('01/01/2024'),
        addedBy: testCMSUser(),
        documents: [
            {
                name: 'Test Question',
                s3URL: 's3://bucketname/key/test1',
                downloadURL: 'https://fake-bucket.s3.amazonaws.com/test',
            },
        ],
        division: 'DMCO',
        responses: [
            {
                id: 'test-question-response-id-1',
                questionID: 'test-question-id-1',
                createdAt: new Date('01/02/2024'),
                addedBy: testStateUser(),
                documents: [
                    {
                        name: 'Test Question Response',
                        s3URL: 's3://bucketname/key/test1response',
                        downloadURL:
                            'https://fake-bucket.s3.amazonaws.com/test1response',
                    },
                ],
            },
        ],
        ...questionData,
    }

    const defaultResponses = [
        {
            id: uuidv4(),
            questionID: question.id,
            //Add 1 day to date, to make sure this date is always after question.createdAt
            createdAt: ((): Date => {
                const responseDate = new Date(question.createdAt)
                return new Date(
                    responseDate.setDate(responseDate.getDate() + 1)
                )
            })(),
            addedBy: testStateUser(),
            documents: [
                {
                    name: 'Test Question Response',
                    s3URL: 's3://bucketname/key/test1',
                    downloadURL: 'https://fake-bucket.s3.amazonaws.com/test',
                },
            ],
        },
    ]

    // If responses are passed in, use that and replace questionIDs, so they match the question.
    question.responses = questionData?.responses
        ? questionData.responses.map((response) => ({
              ...response,
              questionID: question.id,
          }))
        : defaultResponses

    return question as R
}

const mockQuestionAndResponses = (
    question: Partial<ContractQuestionType> & { contractID: string }
): ContractQuestionType => mockCommonQuestionAndResponses(question)

const mockRateQuestionAndResponses = (
    question: Partial<RateQuestionType> & { rateID: string }
): RateQuestionType => mockCommonQuestionAndResponses(question)

export {
    testEmailConfig,
    testStateAnalystsEmails,
    testDuplicateEmailConfig,
    testDuplicateStateAnalystsEmails,
    mockContractRev,
    mockContract,
    mockUnlockedContract,
    testEmailer,
    mockQuestionAndResponses,
    mockRateQuestionAndResponses,
    mockRate,
    testEmailerFromDatabase,
    mockRateRevision,
}
