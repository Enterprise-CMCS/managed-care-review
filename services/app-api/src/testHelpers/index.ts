export {
    assertAnError,
    assertAnErrorExtensions,
    assertAnErrorCode,
} from './gqlAssertions'

export { must } from './assertionHelpers'

export {
    mockInsertContractArgs,
    mockContractRevision,
} from './contractDataMocks'

export {
    mockInsertRateArgs,
    mockRateRevision,
    mockDraftRate,
} from './rateDataMocks'

export { testS3Client } from './s3Helpers'
export { getStateRecord } from './stateHelpers'

export { consoleLogFullData } from './debugHelpers'

export {
    fetchTestRateById,
    submitTestRate,
    unlockTestRate,
    updateTestRate,
} from './gqlRateHelpers'

export {
    createAndSubmitTestContract,
    fetchTestContract,
    approveTestContract,
    updateTestContractDraftRevision,
    createTestContract,
    fetchTestContractWithQuestions,
} from './gqlContractHelpers'

export {
    clearDocMetadata,
    clearMetadataFromContractFormData,
    clearMetadataFromRateFormData,
} from './documentHelpers'

export { mockGqlContractDraftRevisionFormDataInput } from './gqlContractInputMocks'
