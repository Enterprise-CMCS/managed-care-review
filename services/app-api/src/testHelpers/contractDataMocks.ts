import { v4 as uuidv4 } from 'uuid'
import type {
    ContractRevisionTableWithRates,
    ContractTableFullPayload,
} from '../postgres/contractAndRates/prismaFullContractRateHelpers'
import type { StateCodeType } from '@mc-review/submissions'
import type { ContractFormDataType } from '../domain-models'
import { findStatePrograms, type InsertContractArgsType } from '../postgres'
import { must } from './assertionHelpers'

const defaultContractData = () => ({
    id: uuidv4(),
    createdAt: new Date(),
    updatedAt: new Date(),
    mccrsID: null,
    stateCode: 'MN',
    stateNumber: 111,
})

const mockInsertContractArgs = ({
    stateCode = 'MN',
    ...formData
}: {
    stateCode?: StateCodeType
} & Partial<ContractFormDataType>): InsertContractArgsType => {
    const statePrograms = must(findStatePrograms(stateCode))

    return {
        stateCode: stateCode,
        submissionType: formData?.submissionType ?? 'CONTRACT_AND_RATES',
        submissionDescription:
            formData?.submissionDescription ?? 'Contract 1.0',
        contractType: formData?.contractType ?? 'BASE',
        programIDs: formData?.programIDs ?? [statePrograms[0].id],
        populationCovered: formData?.populationCovered ?? 'MEDICAID',
        riskBasedContract: formData?.riskBasedContract ?? false,
        ...formData,
    }
}

const mockContractRevision = (
    contract?: Partial<ContractTableFullPayload>,
    revision?: Partial<ContractRevisionTableWithRates>,
    stateCode: StateCodeType = 'MN'
): ContractRevisionTableWithRates => {
    const statePrograms = must(findStatePrograms(stateCode))

    return {
        id: uuidv4(),
        contract: {
            ...defaultContractData(),
            ...contract,
        },
        relatedSubmisions: [],
        createdAt: new Date(),
        updatedAt: new Date(),
        submitInfo: {
            id: uuidv4(),
            updatedAt: new Date(),
            updatedByID: 'someone',
            updatedReason: 'submit',
            updatedBy: {
                id: 'someone',
                createdAt: new Date(),
                updatedAt: new Date(),
                givenName: 'Bob',
                familyName: 'Law',
                email: 'boblaw@example.com',
                role: 'STATE_USER',
                divisionAssignment: null,
                stateCode: stateCode,
            },
        },
        dsnpContract: null,
        unlockInfo: null,
        contractID: 'contractID',
        submitInfoID: null,
        unlockInfoID: null,
        programIDs: [statePrograms[0].id],
        populationCovered: 'MEDICAID' as const,
        submissionType: 'CONTRACT_ONLY' as const,
        riskBasedContract: false,
        submissionDescription: 'Test',
        stateContacts: [],
        supportingDocuments: [
            {
                id: uuidv4(),
                position: 0,
                contractRevisionID: 'contractRevisionID',
                createdAt: new Date(),
                updatedAt: new Date(),
                name: 'contract supporting doc',
                s3URL: 's3://bucketname/key/test1',
                sha256: '2342fwlkdmwvw',
                dateAdded: new Date(),
            },
            {
                id: uuidv4(),
                position: 1,
                contractRevisionID: 'contractRevisionID',
                createdAt: new Date(),
                updatedAt: new Date(),
                name: 'contract supporting doc 2',
                s3URL: 's3://bucketname/key/test1',
                sha256: '45662342fwlkdmwvw',
                dateAdded: new Date(),
            },
        ],
        contractType: 'BASE',
        contractExecutionStatus: 'EXECUTED',
        contractDocuments: [
            {
                id: uuidv4(),
                position: 0,
                contractRevisionID: 'contractRevisionID',
                createdAt: new Date(),
                updatedAt: new Date(),
                name: 'contract doc',
                s3URL: 's3://bucketname/key/test1',
                sha256: '8984234fwlkdmwvw',
                dateAdded: new Date(),
            },
        ],
        contractDateStart: new Date(Date.UTC(2025, 5, 1)),
        contractDateEnd: new Date(Date.UTC(2026, 4, 30)),
        managedCareEntities: ['MCO'],
        federalAuthorities: ['STATE_PLAN' as const],
        modifiedBenefitsProvided: false,
        modifiedGeoAreaServed: false,
        modifiedMedicaidBeneficiaries: false,
        modifiedRiskSharingStrategy: false,
        modifiedIncentiveArrangements: false,
        modifiedWitholdAgreements: false,
        modifiedStateDirectedPayments: false,
        modifiedPassThroughPayments: false,
        modifiedPaymentsForMentalDiseaseInstitutions: false,
        modifiedMedicalLossRatioStandards: false,
        modifiedOtherFinancialPaymentIncentive: false,
        modifiedEnrollmentProcess: false,
        modifiedGrevienceAndAppeal: false,
        modifiedNetworkAdequacyStandards: true,
        modifiedLengthOfContract: true,
        modifiedNonRiskPaymentArrangements: null,
        inLieuServicesAndSettings: null,
        statutoryRegulatoryAttestation: null,
        statutoryRegulatoryAttestationDescription: null,
        ...revision,
    }
}

export { mockInsertContractArgs, mockContractRevision }
