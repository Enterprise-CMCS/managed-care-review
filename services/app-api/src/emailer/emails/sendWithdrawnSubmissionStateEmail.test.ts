import { mockContract, testEmailConfig } from '../../testHelpers/emailerHelpers'
import { sendWithdrawnSubmissionStateEmail } from './sendWithdrawnSubmissionStateEmail'

const testContract = mockContract({
    id: 'test-contract-123',
    consolidatedStatus: 'WITHDRAWN',
    reviewStatusActions: [
        {
            updatedAt: new Date('2025-04-10'),
            updatedBy: {
                role: 'CMS_USER',
                email: 'zuko@example.com',
                givenName: 'Zuko',
                familyName: 'Hotman',
            },
            updatedReason: 'Test Withdraw',
            dateApprovalReleasedToState: undefined,
            actionType: 'WITHDRAW',
            contractID: 'test-contract-123',
        },
    ],
})

const testRatesForDisplay = [
    {
        id: 'rate-123',
        rateCertificationName: 'rate-name-123',
    },
    {
        id: 'rate-321',
        rateCertificationName: 'rate-name-321',
    },
]

describe('sendWithdrawnSubmissionStateEmail', () => {
    it('State email contains correct toAddresses', async () => {
        const template = await sendWithdrawnSubmissionStateEmail(
            testContract,
            testRatesForDisplay,
            testEmailConfig()
        )

        if (template instanceof Error) {
            throw template
        }

        expect(template.toAddresses).toEqual(
            expect.arrayContaining([
                'contract-state-contact@example.com',
                ...testEmailConfig().devReviewTeamEmails,
            ])
        )
    })

    it('renders overall email for withdrawn contract as expected', async () => {
        const template = await sendWithdrawnSubmissionStateEmail(
            testContract,
            testRatesForDisplay,
            testEmailConfig()
        )

        if (template instanceof Error) {
            throw template
        }

        expect(template.bodyHTML).toMatchSnapshot()
    })
})
