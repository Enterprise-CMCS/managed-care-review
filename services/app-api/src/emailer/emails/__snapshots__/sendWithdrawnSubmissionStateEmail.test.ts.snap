// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sendWithdrawnSubmissionStateEmail > renders overall email for withdrawn contract as expected 1`] = `
"<h1 style="font-weight: normal;">Submission MCR-MN-0004-SNBC was withdrawn by CMS</h1>
<b>Withdrawn by:</b> zuko@example.com<br />
<b>Withdrawn on:</b> 04/09/2025<br />
<b>Reason for withdrawal:</b> Test Withdraw<br />
<b>Related rate certification(s):</b>
<ul>
<li>rate-name-123</li>
<li>rate-name-321</li>
</ul>
<a href="http://localhost/submissions/test-contract-123">View submission in MC-Review</a>"
`;
